File: csharp/medium/range-product-queries-of-powers.cs
IList<int> ProductQueries(int n, int[][] queries) {
        const int mod = 1_000_000_007;
        List<long> bins = new List<long>();
        long rep = 1;

        while (n > 0) {
            if ((n & 1) != 0) {
                bins.Add(rep);
            }
            n >>= 1;
            rep <<= 1;
        }

        int m = bins.Count;
        long[] prefix = new long[m];
        prefix[0] = bins[0] % mod;

        for (int i = 1; i < m; i++) {
            prefix[i] = (prefix[i - 1] * bins[i]) % mod;
        }

        long ModPow(long baseVal, long exp, long modVal) {
            long result = 1;
            baseVal %= modVal;
            while (exp > 0) {
                if ((exp & 1) != 0) result = (result * baseVal) % modVal;
                baseVal = (baseVal * baseVal) % modVal;
                exp >>= 1;
            }
            return result;
        }

        long ModInv(long x) {
            return ModPow(x, mod - 2, mod); // Fermat's little theorem
        }

        List<int> ans = new List<int>();
        foreach (var query in queries) {
            int left = query[0];
            int right = query[1];
            if (left == 0) {
                ans.Add((int)prefix[right]);
            } else {
                ans.Add((int)((prefix[right] * ModInv(prefix[left - 1])) % mod));
            }
        }

        return ans;
    }

File: typescript/medium/range-product-queries-of-powers.ts
function productQueries(n: number, queries: number[][]): number[] {
    const mod = 1_000_000_007;
    let bins: number[] = [];
    let rep = 1;

    while (n > 0) {
        if ((n & 1) !== 0) {
            bins.push(rep);
        }
        n >>= 1;
        rep <<= 1;
    }

    const m = bins.length;
    let prefix: number[] = Array(m).fill(1);
    prefix[0] = bins[0] % mod;

    for (let i = 1; i < m; i++) {
        prefix[i] = (prefix[i - 1] * bins[i]) % mod;
    }

    function modInv(x: number): number {
        return powMod(x, mod - 2, mod); // Fermat's little theorem
    }

    function powMod(base: number, exp: number, mod: number): number {
        let result = 1;
        base %= mod;
        while (exp > 0) {
            if ((exp & 1) !== 0) result = (result * base) % mod;
            base = (base * base) % mod;
            exp >>= 1;
        }
        return result;
    }

    let ans: number[] = [];
    for (let [left, right] of queries) {
        if (left === 0) {
            ans.push(prefix[right]);
        } else {
            ans.push((prefix[right] * modInv(prefix[left - 1])) % mod);
        }
    }

    return ans;
}
